modules/cdc-ext/src/main/java/org/apache/ignite/cdc/kafka/KafkaToIgniteCdcStreamer.java

```diff
@@ -0,0 +1,181 @@
+    private final Thread[] runners;
+
+    /** Appliers. */
+    private final List<KafkaToIgniteCdcStreamerApplier> appliers;
+
+    /**
+     * @param igniteCfg Ignite configuration.
+     * @param kafkaProps Kafka properties.
+     * @param streamerCfg Streamer configuration.
+     */
+    public KafkaToIgniteCdcStreamer(
+        IgniteConfiguration igniteCfg,
+        Properties kafkaProps,
+        KafkaToIgniteCdcStreamerConfiguration streamerCfg
+    ) {
+        this.igniteCfg = igniteCfg;
+        this.kafkaProps = kafkaProps;
+        this.streamerCfg = streamerCfg;
+
+        appliers = new ArrayList<>(streamerCfg.getThreadCount());
+        runners = new Thread[streamerCfg.getThreadCount()];
+
+        if (!kafkaProps.containsKey(ConsumerConfig.GROUP_ID_CONFIG))
+            throw new IllegalArgumentException("Kafka properties don't contains " + ConsumerConfig.GROUP_ID_CONFIG);
+
+        kafkaProps.put(KEY_DESERIALIZER_CLASS_CONFIG, IntegerDeserializer.class.getName());
+        kafkaProps.put(VALUE_DESERIALIZER_CLASS_CONFIG, ByteArrayDeserializer.class.getName());
+    }
+
+    /** {@inheritDoc} */
+    @Override public void run() {
+        try (IgniteEx ign = (IgniteEx)Ignition.start(igniteCfg)) {
+            IgniteLogger log = U.initLogger(igniteCfg, "kafka-ignite-streamer");
+
+            AtomicBoolean closed = new AtomicBoolean();
+
+            Set<Integer> caches = null;
+
+            if (!F.isEmpty(streamerCfg.getCaches())) {
+                caches = streamerCfg.getCaches().stream()
+                    .peek(cache -> Objects.requireNonNull(ign.cache(cache), cache + " not exists!"))
+                    .map(CU::cacheId).collect(Collectors.toSet());
+            }
+
+            int kafkaParts = streamerCfg.getKafkaPartitions();
+            int threadCnt = streamerCfg.getThreadCount();
+            int partPerApplier = kafkaParts / threadCnt;
+
+            for (int i = 0; i < threadCnt; i++) {
+                int from = i * partPerApplier;
+                int to = (i + 1) * partPerApplier;
+
+                if (i == threadCnt - 1)
+                    to = kafkaParts + 1;
+
+                KafkaToIgniteCdcStreamerApplier applier = new KafkaToIgniteCdcStreamerApplier(
+                    ign,
+                    log,
+                    kafkaProps,
+                    streamerCfg.getTopic(),
+                    from,
+                    to,
+                    caches,
+                    streamerCfg.getMaxBatchSize(),
+                    closed
+                );
+
+                appliers.add(applier);
+
+                runners[i] = new Thread(applier, "applier-thread-" + i);
+
+                runners[i].start();
+            }
+
+            try {
+                for (int i = 0; i < threadCnt; i ++)
+                    runners[i].join();
+            }
+            catch (InterruptedException e) {
+                closed.set(true);
+
+                appliers.forEach(U::closeQuiet);
```

Comment:
any reason to call the method which sets `closed.set(true);` line after it set? %)

Reply:
`close` invocation required to call `KafkaConsumer#wakeUp` to cancel current consumer poll request.

Reply:
you performing `closed.set(true);` twice